pyqt-apps/siriushla/as_ps_cycle/pvnames_tree.py

```diff
@@ -0,0 +1,228 @@
+                item = parent.children[val].item
+            else:
+                if not pvals:
+                    item = QTreeWidgetItem(row)
+                    self._leafs.append(item)
+                else:
+                    item = QTreeWidgetItem([val])
+                item.setCheckState(0, Qt.Unchecked)
+                parent.children[val] = PVNameTree._node(item, dict())
+
+            if parent.item is not None:
+                parent.item.addChild(item)
+            parent = parent.children[val]
+            i += 1
+
+        # self.itemInserted.emit()
+
+    def _setup_ui(self):
+        self._add_items()
+        self.expanded.connect(
+            lambda idx: self.resizeColumnToContents(idx.column()))
+
+    def _add_items(self):
+        dlg = QProgressDialog(
+            labelText='Building Tree',
+            minimum=0, maximum=len(self._items), parent=self)
+        t = self.BuildTree(self)
+        # Connect signals
+        t.itemInserted.connect(lambda: dlg.setValue(dlg.value() + 1))
+        t.finished.connect(dlg.close)
+        t.finished.connect(t.deleteLater)
+        # Start
+        t.start()
+        dlg.exec_()
+
+    def _item_checked(self, item, column):
+        if self.check_children and self.check_parent:
+            self.check_parent = False
+            self._check_children(item, column)
+            self.check_parent = True
+            if item.parent():
+                self.check_children = False
+                self._check_parent(item.parent())
+                self.check_children = True
+        elif self.check_children and not self.check_parent:
+            self._check_children(item, column)
+        elif not self.check_children and self.check_parent:
+            if item.parent():
+                self._check_parent(item.parent())
+
+    def _check_children(self, item, column):
+        if column == 0:
+            check_state = item.checkState(0)
+            child_count = item.childCount()
+            for i in range(child_count):
+                item.child(i).setCheckState(0, check_state)
+
+    def _check_parent(self, item):
+        child_count = item.childCount()
+        checked = 0
+        unchecked = 0
+        for i in range(child_count):
+            if item.child(i).checkState(0) == Qt.Checked:
+                checked += 1
+            elif item.child(i).checkState(0) == Qt.Unchecked:
+                unchecked += 1
+
+        if checked == child_count:
+            item.setCheckState(0, Qt.Checked)
+        elif unchecked == child_count:
+            item.setCheckState(0, Qt.Unchecked)
+        else:
+            item.setCheckState(0, Qt.PartiallyChecked)
+
+    def checked_items(self):
+        """Return checked items."""
+        return [item.data(0, Qt.DisplayRole) for item in self._leafs
+                if item.checkState(0) == Qt.Checked]
+
+    def sizeHint(self):
+        """Override sizehint."""
+        return QSize(600, 600)
+
+    def _device_type(self, name):
```

Comment:
I think this method could be exposed (without the `_`) and maybe be called `mag_group`, or something in that line. It could also be decorated with `@staticmethod`.

Reply:
Renamed